### ISSUE TITLE
Load and dump data to json files

### ISSUE THREAD

**Issue by rlaverde** [lektor/lektor#477]
It will be nice to have cli commands for importing/exporting data, something like:

`$ lektor load file.csv`
`
$ lektor dump model --otuput file.csv`

It'll be useful when importing data from other tools, spreadsheets, databases.. etc

- It's  this ok?
- Should data  be imported/exported from csv or json files?

If this gets approved I'm willing to provide a PR for it

**Comment by goanpeca** [comment]
I think this would be really useful, but I am partial to json format, although both csv, json, should be possible :-)

**Comment by Daniel-at-github** [comment]
In other static site generators I have seen it as a plugin, and read the data in it's original format.
The example that I can find now is https://github.com/pyvideo/pyvideo/blob/master/plugins/json_reader.py

**Comment by nixjdm** [comment]
Generally I like this, but I don't have a clear understanding of what you envision, @rlaverde. Do you mean exporting contents files and databags? That would involve adding a translation layer from our unique lr file format to/from  json/csv, right? There might be some interesting choices that come up in how to do that. For databags ini -> csv should be easier.

Also, `lektor load file.csv` reads like a pretty broadly scoped command. Where do you envision this being imported to? Will that be encoded in the csv? CLI options could let you specify what to do with it.

**Comment by goanpeca** [comment]
@nixjdm what we have talked with @rlaverde on the topic is.

Lets assume I have some lektor site that defines some models, users, talks, rooms, you name it. And I have a list of 100+ entries (from somewhere else) and I would like to upload to these models. 

Maybe I already know the relation between models and I have a JSON structured in a way that I can populate the database, after all Lektor has one, albeit a flat-file-based-"simple"-one. I could generate `content.lr` files programmatically (and this is what we have done so far) but there is an api for that, why not use it?

The same goes for the reverse. Lektor is a great solution, but maybe someone needs to migrate to django or flask for ________ reason, so it should be equally easy to dump all the DB content in a "structured" JSON file, or a bunch of csv files.

The idea was not at all about databags, was about the actual content of the DB

If we start with a json file doing a dump of the whole DB, I think it would be clear what we could do in the other direction (load)

**Comment by nixjdm** [comment]
I think I get it.

> If we start with a json file doing a dump of the whole DB, I think it would be clear what we could do in the other direction (load)

Dumping the data sounds more straight forward now. I bet you're right that once that's done seeing how to load it back in would be a lot easier. I imagine we'll end up describing something like a spec for the format of the csv/json data so that it makes some sense in the context of lektor.

**Comment by goanpeca** [comment]
> I imagine we'll end up describing something like a spec for the format of the csv/json data so that it makes some sense in the context of lektor.

Exactly!, so what do you think about this now?

**Comment by nixjdm** [comment]
I like it! I'm still awfully curious how the load direction will work out, but there's got to be a cleanish way of doing it. The data in Lektor's db isn't *that* complex. A spec hopefully won't be that hard to make. It'll certainly have to be documented :)

**Comment by goanpeca** [comment]
Ok @nixjdm I will work on it. This could also be a plugin I guess?, but I feel it should be core :-p

**Comment by nixjdm** [comment]
Yeah, I feel like it should be core as well. I nice addition to the cli. Let me know how it goes!

**Comment by rlaverde** [comment]
> Also,` lektor load file.csv` reads like a pretty broadly scoped command. Where do you envision this being imported to? Will that be encoded in the csv? CLI options could let you specify what to do with it.

Yes, load command will need some options (also dump), maybe something like this:

`lektor load --model <model> --path <path> --format <json,csv> <file name>`

`lektor load --model blog --path /content/blog --format csv blog_dump.csv`

**Comment by torian257x** [comment]
@rlaverde sounds awesome. Btw more common is CSV imho. As I would give sheets to clients to fill them out. 
A convention or configuration would be needed for which field maps to where. Probably just the header row of the csv contains the field names of the model.

Waiting for that PR! 👍

**Comment by torian257x** [comment]
a nice addition would probably be that we can use a plugin in between. E.g. you provide a hook during the import where one can write a plugin. A typical example would be image downloads... say you have a column with a URL to an image or some attachment and want to download it as well and put it in that content folder. Or you know, other things like in my case I would need to get the `<h1>` title content and save it with the URL. I could then even write a plugin to take care of the specific cases, say `if url contains somewebpage.com then put image x.jpg as logo image` 
But that is additional sugar

**Comment by nixjdm** [comment]
It's worth noting, due to #439 and #532, that if possible we should try to export and import in ways that are most painless for transferring data with other DBs and static site generators. Other static site generators do not share Lektor's scheme though, so we can probably only do so much.

**Comment by nixjdm** [comment]
In case anyone wants to take a crack at this a good place to start might be datamodel.py For instance it has several `to_json` methods that can do a lot of the exporting right now. Importing is another matter, but it's a good jumping off point.

For instance, play around in the `lektor dev shell`. Try `pprint.pprint(pad.get('/').datamodel.to_json(pad))`
